### PR TITLE
Docs

### PR DESCRIPTION
**Pull request by waterlink**
refs #189 

This PR adds proper `aruba/cucumber/relish` setup for alive documentation through BDD.

+ few documented pages.

@egonSchiele you probably need access to those publishers I have created on relish:

- https://relishapp.com/contracts-staging
- https://relishapp.com/contracts

For that, can you signup (if you haven't already) on https://relishapp.com/ and send me your relish username (here or by email)?

Relish was chosen because of impression created by `rspec`'s team documentation there. It really looks good and offers nice features.

This PR does not add all the documentation yet, but it would be nice to get this reviewed, code-review-fixed and merged, so that we can go in small increments with this.

How it would look when released can be seen at https://relishapp.com/contracts-staging and https://relishapp.com/contracts-staging/contracts-ruby

Cheers!

### CODE REVIEW

**Comment by waterlink** [comment]
PR is quite big. Might make more sense to review it by one commit.

**Comment by egonSchiele** [comment]
Thanks for doing this @waterlink ! I'll make an account and take a look soon.

**Comment by egonSchiele** [comment]
@waterlink my username is `adit`.

**Comment by waterlink** [comment]
I have added you to these publishers. You should be able to see them at your dashboard at https://relishapp.com/my/publishers

**Comment by waterlink** [comment]
BTW, @egonSchiele Can you take a look at a PR, and how it looks on staging publisher? It is important to merge that, because after that multiple small documentation changes can be done in parallel in their respective `*.feature` files, even by multiple contributors! - which is awesome :)

Anybody else wants to review? /cc @sfcgeorge @nixpulvis

**Comment by waterlink** [comment]
Don't be afraid of `+1537, -0`. It is documentation - pure markdown text and a bit of code examples (not spaghetti code that needs to be understood :D )

File: features/basics/functype.feature [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      end
+
+      Contract String => String
+      def self.greeting(name)
+        "Hello, #{name}"
+      end
+
+      class << self
+        Contract C::Num => String
+        def th(number)
```
Comment (egonSchiele) [inline features/basics/functype.feature]: I would suggest a less complicated function, or just remove this one...two examples is enough.

Comment (egonSchiele) [inline features/basics/functype.feature]: Ah, I guess you want the third example as the singleton example. Maybe `def increment(x); x + 1; end` ?

**Comment by egonSchiele** [comment]
This is really cool @waterlink, thanks for doing it.

File: features/basics/functype.feature (hunk 1 above)
Comment (alex-fedorov) [inline features/basics/functype.feature]: Will change it now.

Comment (alex-fedorov) [inline features/basics/functype.feature]: added commit on top of master: https://github.com/egonSchiele/contracts.ruby/commit/51cfe7cc57ef102074f52bdef06a2301c4542d6b

File: features/basics/functype.feature [hunk 2]
```diff
@@ -0,0 +1,76 @@
+Feature: Fetching contracted function type
+
+  You can use `functype(name)` method for that:
+
+  ```ruby
+  functype(:add) # => "add :: Num, Num => Num"
```
Comment (dg-ratiodata) [inline features/basics/functype.feature]: Is that really something "normal" end users are going to use? Otherwise I would move that to "features/developers/..." or something similar.

**Comment by dg-ratiodata** [comment]
@waterlink I think you're on a very good way. :-)

What I found out while "writing" our documentation:

1. Concentrate on API which is relevant to your users
2. Write the tests in a way you expect your users to use the API
3. Some edge case might be better documented by `Rspec` or something similar and is not exposed to your users

Something I tend to do:
* Combine `cucumber` and `rspec` for documentation - example: https://github.com/cucumber/aruba/blob/master/features/api/command/run.feature, this [step](https://github.com/cucumber/aruba/blob/master/features/api/command/run.feature#L24) will be part of the upcoming `aruba` release. For now you need to copy the relevant lines from [this](https://github.com/cucumber/aruba/blob/master/lib/aruba/cucumber/testing_frameworks.rb) file.

Something like that - not tested.

~~~cucumber
Scenario: No Instance method contract violation
  Given a file named "contract.rb" with:
  """ruby
  require "contracts"
  C = Contracts

  class Example
    include Contracts::Core

    Contract C::Num, C::Num => C::Num
    def add(a, b)
      a + b
    end
  end
  """
  And a file named "spec/instance_method_violation_spec.rb" with:
  """
  require 'contract.rb'

  RSpec.describe "instance method violation" do
    subject { Example.new }

    let(:a) { 1 }
    let(:b) { 2 }

    it { expect(subject.add(a, b)).to eq 3 }
  end
  """
  When I run `rspec`
  Then the specs should all pass

Scenario: Instance method contract violation
  Given a file named "contract.rb" with:
  """ruby
  require "contracts"
  C = Contracts

  class Example
    include Contracts::Core

    Contract C::Num, C::Num => C::Num
    def add(a, b)
      a + b
    end
  end
  """
  And a file named "spec/instance_method_violation_spec.rb" with:
  """
  require 'contract.rb'

  RSpec.describe "instance method violation" do
    subject { Example.new }

    let(:a) { 1 }
    let(:b) { 'a' }

    it { expect { subject.add(a, b) }.to raise_error ParamContractError }
  end
  """
  When I run `rspec`
  Then the specs should all pass
~~~

Using this approach you do not need to do something like `puts XY` but you can use the same "synatx" your users are using.

Hope this makes sense to you.

**Comment by sfcgeorge** [comment]
Oh this is pretty cool, nice!

File: features/basics/functype.feature (hunk 2 above)
Comment (alex-fedorov) [inline features/basics/functype.feature]: Makes sense to move it to `features/developers/functype.feature` I think.

**Comment by alex-fedorov** [comment]
@dg-ratiodata For contract violation errors I wanted to demonstrate exactly how error looks like and what data it has inside. How would you do that with `RSpec` ?

**Comment by dg-ratiodata** [comment]
Oh, ok. But what about something like that:

~~~ruby
let(:error_message) { <<-EOS
    : Contract violation for argument 2 of 2: (ParamContractError)
            Expected: Num,
            Actual: "foo"
            Value guarded in: Example::add
            With Contract: Num, Num => Num
            At: instance_method_violation.rb:8
EOS
}

it { expect { subject.add(a, b) }.to raise_error ParamContractError,  error_message }
~~~

**Comment by waterlink** [comment]
That sounds good. I will try it out. Thanks! :)

--
Best Regards,
Alexey Fedorov,
Sr Ruby, Clojure, Crystal, Golang Developer,
Microservices Backend Engineer,
+49 15757 486 476

2015-09-08 9:27 GMT+02:00 Dennis Günnewig <notifications@github.com>:
